pkg/puller/puller.go

```diff
@@ -324,24 +308,42 @@ func (p *Puller) histSyncWorker(ctx context.Context, peer swarm.Address, bin uin
 			if err := p.addPeerInterval(peer, bin, s, top); err != nil {
 				p.metrics.HistWorkerErrCounter.Inc()
 				p.logger.Error(err, "histSyncWorker could not persist interval for peer", "peer_address", peer)
-				continue
+				return false
 			}
 			loggerV2.Debug("histSyncWorker pulled", "bin", bin, "start", s, "topmost", top, "duration", time.Since(syncStart), "peer_address", peer)
 		}
 
 		if err != nil {
 			p.metrics.HistWorkerErrCounter.Inc()
-			loggerV2.Debug("histSyncWorker interval failed", "peer_address", peer, "bin", bin, "cursor", cur, "start", s, "topmost", top, "err", err)
+			p.logger.Debug("histSyncWorker interval failed", "peer_address", peer, "bin", bin, "cursor", cur, "start", s, "topmost", top, "err", err)
 			if errors.Is(err, context.DeadlineExceeded) {
 				p.logger.Debug("histSyncWorker interval timeout, exiting", "total_duration", time.Since(loopStart), "peer_address", peer, "error", err)
 				err = p.blockLister.Blocklist(peer, histSyncTimeoutBlockList, "sync interval timeout")
 				if err != nil {
 					p.logger.Debug("histSyncWorker timeout disconnect error", "error", err)
 				}
-				return
+				return true
 			}
-			sleep = true
-			continue
+		}
+
+		return false
+	}
+
+	for {
+		p.metrics.HistWorkerIterCounter.Inc()
+
+		select {
+		case <-ctx.Done():
+			loggerV2.Debug("histSyncWorker context cancelled", "peer_address", peer, "bin", bin, "cursor", cur)
+			return
+		case p.activeHistoricalJobs <- struct{}{}:
```

Comment:
Not sure what is the main purpose of the `activeHistoricalJobs` channel?

Reply:
it's to limit the number of sync jobs that are inflight